File: scripts/datagen.js
// Reference implementation of the vector generated-data algorithm and its
// derived digest values. Normative prose lives in README.md ("Generated data").
//
// Usage:
//   node scripts/datagen.js --self-test
//   (also required as a module by scripts/validate.js)

'use strict'

const crypto = require('node:crypto')

// Bytes per chunk when streaming or digesting. A multiple of the 32-byte prng
// block, so chunk boundaries never split a block.
const CHUNK_SIZE = 1024 * 1024

// --- generation -------------------------------------------------------------

// A dataset resolved to the stream it reads from. `base` is the absolute offset
// of the dataset's byte 0 within that stream, which is what makes a $slice free:
// a slice is its parent with a starting offset. The seed/pattern bytes are
// decoded once here, not per chunk.
function resolve (specs, name) {
  const spec = specs[name]
  if (!spec) throw new Error(`unknown dataset: ${name}`)
  if (spec.$prng) {
    return { seed: Buffer.from(spec.$prng.seed, 'utf8'), pat: null, base: 0, length: spec.$prng.size }
  }
  if (spec.$pattern) {
    const d = spec.$pattern
    let pat
    if (d.pattern !== undefined) pat = Buffer.from(d.pattern, 'utf8')
    else if (d.patternBase64 !== undefined) pat = Buffer.from(d.patternBase64, 'base64')
    else throw new Error(`dataset '${name}': neither pattern nor patternBase64`)
    if (pat.length === 0) throw new Error('empty pattern')
    return { seed: null, pat, base: 0, length: d.size }
  }
  if (spec.$slice) {
    const d = spec.$slice
    const parent = specs[d.of]
    if (!parent) throw new Error(`slice '${name}' references unknown dataset '${d.of}'`)
    if (parent.$slice) throw new Error(`slice '${name}' references slice '${d.of}' (chained slices are not allowed)`)
    const src = resolve(specs, d.of) // validates the parent; generates nothing
    if (d.offset > src.length || d.length > src.length - d.offset) {
      throw new Error(`slice '${name}' [${d.offset}, ${d.offset + d.length}) exceeds '${d.of}' size ${src.length}`)
    }
    return { seed: src.seed, pat: src.pat, base: src.base + d.offset, length: d.length }
  }
  throw new Error(`unknown data kind: ${JSON.stringify(Object.keys(spec))}`)
}

// Written so no intermediate can overflow: never `offset + length > size`.
function checkRange (name, size, offset, length) {
  if (!Number.isInteger(offset) || !Number.isInteger(length) || offset < 0 || length < 0) {
    throw new Error(`invalid range [${offset}, ${length}) for dataset '${name}'`)
  }
  if (offset > size || length > size - offset) {
    throw new Error(`range [${offset}, ${offset + length}) exceeds dataset '${name}' size ${size}`)
  }
}

// Write `n` bytes of `src`, starting at `offset` within the dataset, into `dst`.
// The only windowing code; every public entry point goes through it.
function readInto (src, offset, dst, n) {
  if (n === 0) return // (abs + n - 1) would underflow below
  const abs = src.base + offset

  if (src.pat) {
    // byte N of the stream is pattern[N % L], so a range starts at phase abs % L
    const L = src.pat.length
    const first = Math.min(n, L)
    const phase = abs % L
    for (let k = 0; k < first; k++) dst[k] = src.pat[(phase + k) % L]
    for (let filled = first; filled < n;) {
      const m = Math.min(filled, n - filled)
      dst.copy(dst, filled, 0, m)
      filled += m
    }
    return
  }

  // block(i) = SHA256(UTF8(seed) || BE64(i)); stream = block(0) || block(1) || ...
  const counter = Buffer.alloc(8)
  const lastBlk = Math.floor((abs + n - 1) / 32)
  for (let i = Math.floor(abs / 32); i <= lastBlk; i++) {
    counter.writeBigUInt64BE(BigInt(i))
    const block = crypto.createHash('sha256').update(src.seed).update(counter).digest()
    const blkStart = i * 32
    const lo = Math.max(abs, blkStart) - blkStart // head trim, nonzero on the first block only
    const hi = Math.min(abs + n, blkStart + 32) - blkStart // clamped to the range end, not the dataset size
    block.copy(dst, blkStart + lo - abs, lo, hi)
  }
}

// Materialize one named dataset from a vector's `data` map.
function generate (specs, name) {
  const src = resolve(specs, name)
  const out = Buffer.alloc(src.length)
  readInto(src, 0, out, src.length)
  return out
}

// Materialize `[offset, offset+length)` of a dataset without materializing the rest.
function generateRange (specs, name, offset, length) {
  const src = resolve(specs, name)
  checkRange(name, src.length, offset, length)
  const out = Buffer.alloc(length)
  readInto(src, offset, out, length)
  return out
}

// A bounded-memory byte stream over a dataset (or a range of one). Every chunk is
// `chunkSize` bytes except the last; a zero-length range yields no chunks.
function generateStream (specs, name, opts = {}) {
  const src = resolve(specs, name)
  const offset = opts.offset ?? 0
  const length = opts.length ?? (Number.isInteger(offset) && offset >= 0 ? src.length - offset : 0)
  checkRange(name, src.length, offset, length)
  const chunkSize = opts.chunkSize ?? CHUNK_SIZE
  if (!Number.isInteger(chunkSize) || chunkSize <= 0) {
    throw new Error(`invalid chunkSize ${chunkSize} for dataset '${name}'`)
  }
  let pos = 0
  return new ReadableStream({
    pull (controller) {
      if (pos >= length) {
        controller.close()
        return
      }
      const n = Math.min(chunkSize, length - pos)
      const chunk = Buffer.alloc(n)
      readInto(src, offset + pos, chunk, n)
      pos += n
      controller.enqueue(chunk)
    }
  })
}

// The dataset's declared length in bytes, without generating it.
function dataSize (specs, name) {
  return resolve(specs, name).length
}

// --- checksums ---------------------------------------------------------------

function makeCrc32Table (poly) {
  const table = new Uint32Array(256)
  for (let n = 0; n < 256; n++) {
    let c = n
    for (let k = 0; k < 8; k++) c = c & 1 ? (c >>> 1) ^ poly : c >>> 1
    table[n] = c >>> 0
  }
  return table
}

const CRC32_TABLE = makeCrc32Table(0xEDB88320)
const CRC32C_TABLE = makeCrc32Table(0x82F63B78)

// The update functions carry the raw register, so a digest can be accumulated
// across chunks: seed it with the all-ones init once, xor it out once at the end.
function crc32Update (c, buf, table) {
  for (let i = 0; i < buf.length; i++) c = table[(c ^ buf[i]) & 0xFF] ^ (c >>> 8)
  return c >>> 0
}

function crc32 (buf, table) {
  return (crc32Update(0xFFFFFFFF, buf, table) ^ 0xFFFFFFFF) >>> 0
}

// CRC-64/NVME: reflected poly 0x9A6C9329AC4BC9B5, init/xorout all-ones.
const CRC64_TABLE = (() => {
  const poly = 0x9A6C9329AC4BC9B5n
  const table = new BigUint64Array(256)
  for (let n = 0; n < 256; n++) {
    let c = BigInt(n)
    for (let k = 0; k < 8; k++) c = c & 1n ? (c >> 1n) ^ poly : c >> 1n
    table[n] = c
  }
  return table
})()

function crc64Update (c, buf) {
  for (let i = 0; i < buf.length; i++) {
    c = CRC64_TABLE[Number((c ^ BigInt(buf[i])) & 0xFFn)] ^ (c >> 8n)
  }
  return c
}

function crc64nvme (buf) {
  return crc64Update(0xFFFFFFFFFFFFFFFFn, buf) ^ 0xFFFFFFFFFFFFFFFFn
}

function u32ToBase64 (v) {
  const b = Buffer.alloc(4)
  b.writeUInt32BE(v)
  return b.toString('base64')
}

function u64ToBase64 (v) {
  const b = Buffer.alloc(8)
  b.writeBigUInt64BE(v)
  return b.toString('base64')
}

// --- derived values (the ${data.<name>.<field>} placeholders) ----------------

const DERIVED_FIELDS = [
  'size', 'md5', 'etag', 'sha256', 'sha256B64', 'sha1B64', 'crc32B64', 'crc32cB64', 'crc64nvmeB64'
]

// Digests are computed over chunks so peak memory is one chunk, whatever the
// dataset size. `size` reads no bytes at all. Note this bounds memory, not time:
// every field re-reads the dataset, so do not loop DERIVED_FIELDS over a
// multi-gigabyte dataset.
function derivedChunked (specs, name, field, chunkSize) {
  const src = resolve(specs, name)
  if (field === 'size') return String(src.length)

  let hash = null
  let table = null
  let crc = 0xFFFFFFFF
  let crc64 = 0xFFFFFFFFFFFFFFFFn
  switch (field) {
    case 'md5': case 'etag': hash = crypto.createHash('md5'); break
    case 'sha256': case 'sha256B64': hash = crypto.createHash('sha256'); break
    case 'sha1B64': hash = crypto.createHash('sha1'); break
    case 'crc32B64': table = CRC32_TABLE; break
    case 'crc32cB64': table = CRC32C_TABLE; break
    case 'crc64nvmeB64': break
    default: throw new Error(`unknown derived data field: ${field}`)
  }

  const buf = Buffer.alloc(Math.min(chunkSize, src.length) || 1)
  for (let pos = 0; pos < src.length; pos += chunkSize) {
    const n = Math.min(chunkSize, src.length - pos)
    const chunk = buf.subarray(0, n)
    readInto(src, pos, chunk, n)
    if (hash) hash.update(chunk)
    else if (table) crc = crc32Update(crc, chunk, table)
    else crc64 = crc64Update(crc64, chunk)
  }

  switch (field) {
    case 'md5': return hash.digest('hex')
    case 'etag': return `"${hash.digest('hex')}"`
    case 'sha256': return hash.digest('hex')
    case 'sha256B64': return hash.digest('base64')
    case 'sha1B64': return hash.digest('base64')
    case 'crc32B64': case 'crc32cB64': return u32ToBase64((crc ^ 0xFFFFFFFF) >>> 0)
    default: return u64ToBase64(crc64 ^ 0xFFFFFFFFFFFFFFFFn)
  }
}

function derived (specs, name, field) {
  return derivedChunked(specs, name, field, CHUNK_SIZE)
}

module.exports = {
  generate,
  generateRange,
  generateStream,
  dataSize,
  derived,
  DERIVED_FIELDS,
  CHUNK_SIZE,
  __derivedChunked: derivedChunked
}

// --- self-test ----------------------------------------------------------------

async function selfTest () {
  const assert = require('node:assert')

  // prng blocks cross-checked against `printf 'test\0...' | shasum -a 256`
  const specs = {
    t96: { $prng: { seed: 'test', size: 96 } },
    t40: { $prng: { seed: 'test', size: 40 } },
    t32: { $prng: { seed: 'test', size: 32 } },
    t10: { $prng: { seed: 'test', size: 10 } },
    aaa: { $pattern: { pattern: 'A', size: 5 } },
    abc: { $pattern: { pattern: 'abc', size: 8 } },
    abc100: { $pattern: { pattern: 'abc', size: 100 } },
    bin: { $pattern: { patternBase64: '3q2+7w==', size: 6 } },
    zero: { $pattern: { pattern: 'A', size: 0 } },
    sl: { $slice: { of: 't40', offset: 30, length: 6 } },
    psl: { $slice: { of: 'abc100', offset: 7, length: 20 } }
  }
  const block0 = 'b8cc3d1fcf7818feab07f224263256110eeb3b576a94ef8e7e439b48fc77998b'
  const block1 = '64a3a04c326aae7efd121f8468df1ac90ead2ece1e952353903cbcb6ae47618d'
  const stream40 = block0 + block1.slice(0, 16)
  assert.strictEqual(generate(specs, 't32').toString('hex'), block0)
  assert.strictEqual(generate(specs, 't40').toString('hex'), stream40)
  assert.strictEqual(generate(specs, 't10').toString('hex'), block0.slice(0, 20))
  assert.strictEqual(generate(specs, 'aaa').toString('utf8'), 'AAAAA')
  assert.strictEqual(generate(specs, 'abc').toString('utf8'), 'abcabcab')
  assert.strictEqual(generate(specs, 'bin').toString('hex'), 'deadbeefdead')
  assert.strictEqual(generate(specs, 'sl').toString('hex'), stream40.slice(60, 72))
  assert.strictEqual(generate(specs, 'zero').length, 0)

  // Ranged reads: every expectation is a window of the pinned block hashes.
  const hexRange = (n, o, l) => generateRange(specs, n, o, l).toString('hex')
  assert.strictEqual(hexRange('t40', 30, 6), stream40.slice(60, 72)) // crosses the block boundary
  assert.strictEqual(hexRange('t40', 24, 16), stream40.slice(48, 80)) // starts mid-block, crosses
  assert.strictEqual(hexRange('t40', 36, 4), stream40.slice(72, 80)) // trailing partial block
  assert.strictEqual(hexRange('t96', 0, 32), block0)
  assert.strictEqual(hexRange('t96', 32, 32), block1) // seek to an exact block boundary
  assert.strictEqual(hexRange('sl', 2, 3), stream40.slice(64, 70)) // ranged read of a slice
  assert.strictEqual(generateRange(specs, 'abc', 4, 3).toString('utf8'), 'bca') // mid-phase
  assert.strictEqual(generateRange(specs, 'abc100', 98, 2).toString('utf8'), 'ca') // tail, phase 98 % 3 == 2
  // A slice of a pattern starts mid-period: a port that dropped the slice's base
  // offset would return 'abc' here and still pass every whole-dataset check.
  assert.strictEqual(generateRange(specs, 'psl', 0, 3).toString('utf8'), 'bca')
  assert.strictEqual(generate(specs, 'psl').toString('utf8'), 'bcabcabcabcabcabcabc')
  assert.strictEqual(generateRange(specs, 't40', 40, 0).length, 0)
  assert.strictEqual(dataSize(specs, 'sl'), 6)

  // A range is exactly the corresponding window of the whole dataset. generate()
  // walks from zero; a range divides to find its start block and phase, so these
  // are genuinely different code paths.
  for (const name of ['t96', 't40', 't32', 't10', 'aaa', 'abc', 'abc100', 'bin', 'zero', 'sl', 'psl']) {
    const full = generate(specs, name)
    for (let o = 0; o <= full.length; o++) {
      for (let l = 0; l <= full.length - o; l++) {
        assert.deepStrictEqual(generateRange(specs, name, o, l), full.subarray(o, o + l), `${name} [${o}, ${o + l})`)
      }
    }
  }

  // Streaming concatenates to the same bytes, in chunks of the requested size.
  for (const chunkSize of [1, 7, 32, 1000]) {
    for (const name of ['t96', 't40', 'abc100', 'zero', 'sl', 'psl']) {
      const full = generate(specs, name)
      const sizes = []
      const parts = []
      for await (const chunk of generateStream(specs, name, { chunkSize })) {
        sizes.push(chunk.length)
        parts.push(chunk)
      }
      assert.deepStrictEqual(Buffer.concat(parts), full, `stream ${name}/${chunkSize}`)
      assert.ok(sizes.every(s => s > 0 && s <= chunkSize), `stream ${name}/${chunkSize} chunk sizes`)
      assert.strictEqual(sizes.length, Math.ceil(full.length / chunkSize))
    }
  }

  // Errors: spec validation runs before any byte work, even for an empty range.
  const bad = { ...specs, chain: { $slice: { of: 'sl', offset: 0, length: 1 } }, over: { $slice: { of: 't10', offset: 8, length: 8 } }, nopat: { $pattern: { size: 4 } } }
  assert.throws(() => generate(bad, 'nope'), /unknown dataset/)
  assert.throws(() => generate(bad, 'chain'), /chained slices/)
  assert.throws(() => generate(bad, 'over'), /exceeds/)
  assert.throws(() => generate(bad, 'nopat'), /neither pattern nor patternBase64/)
  assert.throws(() => generateRange(bad, 'chain', 0, 0), /chained slices/)
  assert.throws(() => generateRange(bad, 't10', 8, 8), /exceeds/)
  assert.throws(() => generateRange(bad, 't40', 41, 0), /exceeds/)
  assert.throws(() => generateRange(bad, 't40', 0, 41), /exceeds/)
  assert.throws(() => generateRange(bad, 'sl', 4, 4), /exceeds/)
  assert.throws(() => generateRange(bad, 't40', -1, 1), /invalid range/)
  assert.throws(() => generateStream(bad, 'over'), /exceeds/)

  // Standard CRC check values for the ASCII string "123456789".
  const check = Buffer.from('123456789', 'ascii')
  assert.strictEqual(crc32(check, CRC32_TABLE).toString(16), 'cbf43926')
  assert.strictEqual(crc32(check, CRC32C_TABLE).toString(16), 'e3069283')
  assert.strictEqual(crc64nvme(check).toString(16), 'ae8b14860a799888')

  // Derived values: MD5("AAAAA") etc. computed independently.
  assert.strictEqual(derived(specs, 'aaa', 'md5'), 'f6a6263167c92de8644ac998b3c4e4d1')
  assert.strictEqual(derived(specs, 'aaa', 'etag'), '"f6a6263167c92de8644ac998b3c4e4d1"')
  assert.strictEqual(derived(specs, 'aaa', 'size'), '5')
  assert.strictEqual(derived(specs, 'zero', 'md5'), 'd41d8cd98f00b204e9800998ecf8427e')
  assert.strictEqual(derived(specs, 'zero', 'size'), '0')
  for (const f of DERIVED_FIELDS) derived(specs, 'sl', f) // all fields derivable on slices

  // A digest must not depend on how the dataset was cut into chunks. chunkSize 1
  // makes every byte a boundary, which is what catches a per-chunk xor-out.
  const chk = { ...specs, check: { $pattern: { pattern: '123456789', size: 9 } } }
  for (const name of ['zero', 'check', 'aaa', 'abc100', 't96', 'sl', 'psl', 'bin']) {
    for (const f of DERIVED_FIELDS) {
      const want = derived(chk, name, f)
      for (const chunkSize of [1, 7, 32, 1000, CHUNK_SIZE]) {
        assert.strictEqual(derivedChunked(chk, name, f, chunkSize), want, `${name}.${f}/${chunkSize}`)
      }
    }
  }
  assert.strictEqual(derivedChunked(chk, 'check', 'crc32B64', 1), u32ToBase64(0xcbf43926))

  console.log('datagen self-test: OK')
}

if (require.main === module) {
  if (process.argv.includes('--self-test')) {
    selfTest().catch(e => {
      console.error(e)
      process.exit(1)
    })
  } else {
    console.error('usage: node scripts/datagen.js --self-test')
    process.exit(1)
  }
}

File: packages/python/src/cloud_portable_s3vectors/datagen.py
"""Python port of the vector generated-data reference.

Materializes a vector's named datasets and computes the derived digest strings
that ``${data.<name>.<field>}`` placeholders resolve to. Stdlib-only: CRC-32C
and CRC-64/NVME use pure-Python tables (slow on large datasets; the digest
algorithms are size-independent).

Datasets are seekable: :func:`generate_range` and :func:`generate_stream` read a
window without materializing the rest, and :func:`derived` digests in chunks, so
a multi-gigabyte dataset needs only a chunk of memory.
"""

from __future__ import annotations

import base64
import hashlib
import struct
import zlib
from collections.abc import Iterator, Mapping
from typing import Any

__all__ = [
    "DERIVED_FIELDS", "CHUNK_SIZE", "generate", "generate_range",
    "generate_stream", "data_size", "derived",
]

#: The fields available as ``${data.<name>.<field>}`` placeholders.
DERIVED_FIELDS: tuple[str, ...] = (
    "size", "md5", "etag", "sha256", "sha256B64", "sha1B64",
    "crc32B64", "crc32cB64", "crc64nvmeB64",
)

#: Bytes per chunk when streaming or digesting. A multiple of the 32-byte prng block.
CHUNK_SIZE: int = 1 << 20


class _Source:
    """A dataset resolved to the stream it reads from.

    ``base`` is the absolute offset of the dataset's byte 0 within that stream,
    which is what makes a ``$slice`` free: a slice is its parent with a starting
    offset. Seed and pattern bytes are decoded once, here, not per chunk.
    """

    __slots__ = ("seed", "pat", "base", "length")

    def __init__(self, seed: bytes | None, pat: bytes | None, base: int, length: int) -> None:
        self.seed = seed
        self.pat = pat
        self.base = base
        self.length = length


def _resolve(specs: Mapping[str, Mapping[str, Any]], name: str) -> _Source:
    spec = specs.get(name)
    if spec is None:
        raise KeyError(f"unknown dataset: {name}")
    if "$prng" in spec:
        d = spec["$prng"]
        return _Source(d["seed"].encode("utf-8"), None, 0, d["size"])
    if "$pattern" in spec:
        d = spec["$pattern"]
        if "pattern" in d:
            pat = d["pattern"].encode("utf-8")
        elif "patternBase64" in d:
            pat = base64.b64decode(d["patternBase64"])
        else:
            raise KeyError(f"dataset {name!r}: neither pattern nor patternBase64")
        if not pat:
            raise ValueError("empty pattern")
        return _Source(None, pat, 0, d["size"])
    if "$slice" in spec:
        d = spec["$slice"]
        parent = specs.get(d["of"])
        if parent is None:
            raise KeyError(f"slice {name!r} references unknown dataset {d['of']!r}")
        if "$slice" in parent:
            raise ValueError(f"slice {name!r} references slice {d['of']!r} (chained slices are not allowed)")
        src = _resolve(specs, d["of"])  # validates the parent; generates nothing
        offset, length = d["offset"], d["length"]
        if offset > src.length or length > src.length - offset:
            raise ValueError(
                f"slice {name!r} [{offset}, {offset + length}) exceeds {d['of']!r} size {src.length}"
            )
        return _Source(src.seed, src.pat, src.base + offset, length)
    raise ValueError(f"unknown data kind: {sorted(spec)}")


def _check_range(name: str, size: int, offset: int, length: int) -> None:
    # Written so no intermediate can overflow: never ``offset + length > size``.
    if not isinstance(offset, int) or not isinstance(length, int) or isinstance(offset, bool) or isinstance(length, bool):
        raise ValueError(f"invalid range offset {offset!r} length {length!r} for dataset {name!r}")
    if offset < 0 or length < 0:
        raise ValueError(f"invalid range offset {offset} length {length} for dataset {name!r}")
    if offset > size or length > size - offset:
        raise ValueError(f"range [{offset}, {offset + length}) exceeds dataset {name!r} size {size}")


def _read_range(src: _Source, offset: int, n: int) -> bytes:
    """Produce ``n`` bytes of ``src`` starting at ``offset`` within the dataset.

    The only windowing code; every public entry point goes through it.
    """
    if n == 0:  # (abs + n - 1) would go negative below
        return b""
    abs_off = src.base + offset

    if src.pat is not None:
        # byte N of the stream is pattern[N % L], so a range starts at phase abs % L
        pat = src.pat
        L = len(pat)
        phase = abs_off % L
        rotated = pat[phase:] + pat[:phase]
        return (rotated * (n // L + 1))[:n]

    # block(i) = SHA256(UTF8(seed) || BE64(i)); stream = block(0) || block(1) || ...
    # Filled in place through a memoryview: accumulating one bytes object per
    # 32-byte block costs several times the payload in peak memory.
    seeded = hashlib.sha256(src.seed)
    out = bytearray(n)
    mv = memoryview(out)
    for i in range(abs_off // 32, (abs_off + n - 1) // 32 + 1):
        h = seeded.copy()
        h.update(struct.pack(">Q", i))
        block = h.digest()
        blk_start = i * 32
        lo = max(abs_off, blk_start) - blk_start  # head trim, nonzero on the first block only
        hi = min(abs_off + n, blk_start + 32) - blk_start  # clamped to the range end, not the dataset size
        at = blk_start + lo - abs_off
        if hi - lo == 32:
            mv[at:at + 32] = block  # whole block, no slice of the digest
        else:
            mv[at:at + (hi - lo)] = block[lo:hi]
    mv.release()
    return bytes(out)


def generate(specs: Mapping[str, Mapping[str, Any]], name: str) -> bytes:
    """Materialize one named dataset from a vector's ``data`` map."""
    src = _resolve(specs, name)
    return _read_range(src, 0, src.length)


def generate_range(
    specs: Mapping[str, Mapping[str, Any]], name: str, offset: int, length: int
) -> bytes:
    """Materialize ``[offset, offset+length)`` without materializing the rest."""
    src = _resolve(specs, name)
    _check_range(name, src.length, offset, length)
    return _read_range(src, offset, length)


def generate_stream(
    specs: Mapping[str, Mapping[str, Any]],
    name: str,
    *,
    offset: int = 0,
    length: int | None = None,
    chunk_size: int = CHUNK_SIZE,
) -> Iterator[bytes]:
    """Iterate a dataset (or a range of one) in ``chunk_size`` pieces.

    Every chunk is ``chunk_size`` bytes except the last; a zero-length range
    yields no chunks. This function deliberately contains no ``yield``: a
    generator function would defer every check below to the first ``next()``,
    making Python the only port that validates lazily.
    """
    src = _resolve(specs, name)
    if length is None:
        _check_range(name, src.length, offset, 0)
        length = src.length - offset
    _check_range(name, src.length, offset, length)
    if not isinstance(chunk_size, int) or isinstance(chunk_size, bool) or chunk_size <= 0:
        raise ValueError(f"invalid chunk_size {chunk_size!r} for dataset {name!r}")
    return _iter_range(src, offset, length, chunk_size)


def _iter_range(src: _Source, offset: int, length: int, chunk_size: int) -> Iterator[bytes]:
    pos = 0
    while pos < length:
        n = min(chunk_size, length - pos)
        yield _read_range(src, offset + pos, n)
        pos += n


def data_size(specs: Mapping[str, Mapping[str, Any]], name: str) -> int:
    """The dataset's declared length in bytes, without generating it."""
    return _resolve(specs, name).length


def _make_crc_table(poly: int, width: int) -> list[int]:
    mask = (1 << width) - 1
    table = []
    for n in range(256):
        c = n
        for _ in range(8):
            c = (c >> 1) ^ poly if c & 1 else c >> 1
        table.append(c & mask)
    return table


_CRC32C_TABLE = _make_crc_table(0x82F63B78, 32)
# CRC-64/NVME: reflected poly, init/xorout all-ones.
_CRC64NVME_TABLE = _make_crc_table(0x9A6C9329AC4BC9B5, 64)


def _crc_update(table: list[int], mask: int, reg: int, data: bytes) -> int:
    """Fold ``data`` into a raw CRC register.

    Takes and returns the *raw* register, so a digest can be accumulated across
    chunks: seed it with the all-ones init once and xor it out once at the end.
    (``zlib.crc32`` below uses the other convention — it takes and returns the
    *finalized* value and chains directly from 0.)
    """
    for b in data:
        reg = table[(reg ^ b) & 0xFF] ^ (reg >> 8)
    return reg & mask


def _derived_chunked(
    specs: Mapping[str, Mapping[str, Any]], name: str, field: str, chunk_size: int
) -> str:
    """Digest a dataset in chunks, so peak memory is one chunk at any size.

    This bounds memory, not time: every field re-reads the dataset, so do not
    loop over DERIVED_FIELDS for a multi-gigabyte dataset. ``size`` reads nothing.
    """
    src = _resolve(specs, name)
    if field == "size":
        return str(src.length)

    hasher = None
    table = mask = None
    reg = 0
    crc32_val = 0
    if field in ("md5", "etag"):
        hasher = hashlib.md5()
    elif field in ("sha256", "sha256B64"):
        hasher = hashlib.sha256()
    elif field == "sha1B64":
        hasher = hashlib.sha1()
    elif field == "crc32B64":
        pass
    elif field == "crc32cB64":
        table, mask, reg = _CRC32C_TABLE, 0xFFFFFFFF, 0xFFFFFFFF
    elif field == "crc64nvmeB64":
        table, mask, reg = _CRC64NVME_TABLE, 0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF
    else:
        raise KeyError(f"unknown derived data field: {field}")

    pos = 0
    while pos < src.length:
        n = min(chunk_size, src.length - pos)
        chunk = _read_range(src, pos, n)
        if hasher is not None:
            hasher.update(chunk)
        elif table is not None:
            reg = _crc_update(table, mask, reg, chunk)
        else:
            crc32_val = zlib.crc32(chunk, crc32_val)
        pos += n

    if field == "md5":
        return hasher.hexdigest()
    if field == "etag":
        return f'"{hasher.hexdigest()}"'
    if field == "sha256":
        return hasher.hexdigest()
    if field == "sha256B64":
        return base64.b64encode(hasher.digest()).decode("ascii")
    if field == "sha1B64":
        return base64.b64encode(hasher.digest()).decode("ascii")
    if field == "crc32B64":
        return base64.b64encode(struct.pack(">I", crc32_val & 0xFFFFFFFF)).decode("ascii")
    if field == "crc32cB64":
        return base64.b64encode(struct.pack(">I", reg ^ mask)).decode("ascii")
    return base64.b64encode(struct.pack(">Q", reg ^ mask)).decode("ascii")


def derived(specs: Mapping[str, Mapping[str, Any]], name: str, field: str) -> str:
    """Compute the string a ``${data.<name>.<field>}`` placeholder resolves to."""
    return _derived_chunked(specs, name, field, CHUNK_SIZE)
